src/lib/process/index.ts

```diff
@@ -50,29 +52,44 @@ export const createProcess = async (
   dryRun: boolean = false,
   options: Polkadot.Options = defaultOptions
 ): Promise<Process.Result> => {
-  const program: Process.Program = validate(userProgram)
-  const processId = utf8ToHex(name, Constants.PROCESS_ID_LENGTH)
-  const polkadot: Polkadot.Polkadot = await createNodeApi(options)
-  const expectedVersion: number = (await getVersion(polkadot, processId)) + 1
+  try {
 
-  if (version !== expectedVersion) {
-    throw new VersionError(`Version: ${version} must be incremented to: ${expectedVersion}`)
-  }
+    const program: Process.Program = validate(userProgram)
+    const processId = utf8ToHex(name, Constants.PROCESS_ID_LENGTH)
+    const polkadot: Polkadot.Polkadot = await createNodeApi(options)
+    const currentVersion: number = await getVersion(polkadot, processId)
+    const expectedVersion: number = currentVersion + 1
 
-  if (dryRun)
-    return {
-      process: null,
-      message: 'Dry run: transaction has not been created',
-      name,
-      version: expectedVersion,
-      program,
+    if (version > expectedVersion || version < currentVersion)
+      throw new VersionError(version, expectedVersion, name)
+
+    if (version === currentVersion) {
+      const process = await getProcess(polkadot, processId, version)
+
+      if (program.length !== process.program.length) 
+        throw new ProgramError('existing: programs are different lengths', process) 
+
+      if (!program.every((step, i) => textify(step) === textify(process.program[i])))
+        throw new ProgramError('existing: program steps did not match', process) 
+
+      throw new ProgramError(`Process ${name} is already created.`, process)
     }
 
-  const process: Process.Payload = await createProcessTransaction(polkadot, processId, program, options)
+    if (dryRun)
+      return {
+        process: null,
+        message: 'Dry run: transaction has not been created',
+        name,
+        version: expectedVersion,
+        program,
+      }
 
-  return {
-    message: `Transaction for new process ${name} has been successfully submitted`,
-    process,
+    return {
+      message: `Transaction for new process ${name} has been successfully submitted`,
+      process: await createProcessTransaction(polkadot, processId, program, options),
+    }
+  } catch (err) {
+    return err
   }
```

Comment:
This still doesn't address the question. Now we're returning `unknown` when the expected return is `Process.Result`. Why aren't we just letting the exception propagate?

Reply:
Mainly for multiple processes, we would need to catch. I can see two options here:

`loadProcess` handles try/catch - will be similar due to typescript and try/catch. To my knowledge I would leave `e: any` and then assert each error using `instanceOf` (have a tool like middleware)

In this case Try {} block is what type inherit as you suggested earlier we could wrap somewhere else but there still be e: any or (e) if we want to try/catch. Another solution I can think of is to return Promise.reject() which would cover a type, but then logic would most likely need to change.

Reply:
I understand @n3op2. My point is more about getting value from the return type and not abusing the language. Why not return the errors instead of throwing and then add the error types as part of a type union to the return signature?

Reply:
@mattdean-digicatapult i got to say I like it what I hear, but we still need to handle exception in order to allow load multiple processes.

First idea that comes into my mind is to return `new Promise` instead of try/catch then we can have strongly typed -> https://github.com/digicatapult/dscp-process-management/blob/main/src/index.ts#L102 

Going to have a bit of read how other people address exceptions in ts

Reply:
@mattdean-digicatapult as discussed in the standup I've reverted back to `try/catch` and pushed. Excluded `union` types

Reply:
Can we still add the error types to the return type as that's what you're doing? It should just be changing `Promise<Process.Result>` to `Promise<Process.Result | Error>`

Reply:
it has already been added to the result

Reply:
Thanks for insights. This should have been updated now as per your suggestion.